### ISSUE TITLE
Referring to a GAT

### ISSUE THREAD

**Issue by vigna** [danielhenrymantilla/lending-iterator.rs#13]
This code won't compile:
```rust
#[::nougat::gat]
impl<I: GraphIterator> LendingIterator for Adapter<I> {
    type Item<'a> = (usize, <I as GraphIterator>::Successors<'a>)
    where Self: 'a;

    fn next(&mut self) -> Option<Self::Item<'_>> {
        self.0.next_inner()
    }
}
```
The problem is that apparently the macro fiddles also with `::Sucessors`, while it should just leave it alone. How can we achieve that?

(I realize it is not a compilable snippet, but I think the problem we have is clear—I can produce a compilable example if you need it.)

**Comment by danielhenrymantilla** [comment]
Thanks for the report! While I think about a fix or re-design[^1] to palliate this issue, FWIW, you should be able to work around the issue by defining a helper type alias:

```rs
type GraphIteratorSuccessors<'next, I> = <I as GraphIterator>::Successors<'next>;

#[::nougat::gat]
impl<I : GraphIterator> LendingIterator for Adapter<I> {
    type Item<'next> = (usize, GraphIteratorSuccessors<'next, I>)
    where
        Self : 'next,
    ;

    fn next(&mut self) -> Option<Self::Item<'_>> {
        self.0.next_inner()
    }
}
```

[^1]: I'd like for the next release of `lending-iterator` not to need `nougat`

**Comment by vigna** [comment]
Thanks for the answer! Next question :), how do I make this work:
```rust
pub fn extend<'a, 'b, I: Iterator<Item = usize>, L: LendingIterator<Item<'b> = (usize, I)>>(
&mut self,
        iter_nodes: &'b mut L,
    ){}
```
as the compiler complains that `Item` is not an associated type. I've read your note about using `Item<'_,I>` but I need to constrain the type of the returned item instead (note that the answer might obvious, I've been programming Rust since just six months).

**Comment by danielhenrymantilla** [comment]
So, the main solution here is to be using `#[apply(Gat)]` on the `fn`. The issue, then, is that `LendingIterator` (and nougat) involve `for<'next>` kind of bounds, and your `'b`-specific extra bound then makes it run into a limitation of Rust.

There are then two workarounds:
  - either you write the bound as valid for all lifetimes (while this seems nice, in practice, it will only be usable if the `LendingIterator` isn't truly lending / if `I` does not depend on `'c`)
  - to use a trait "type eq" bound rather than a direct `=` bound, since this allows using an extra line/clause.

```rs
#[apply(Gat!)]
pub fn extend<'a, 'b, I, L>(
    &mut self,
    iter_nodes: &'b mut L,
)
where
    I : Iterator<Item = usize>,
    L : for<'c> LendingIterator<Item<'c> = (usize, I)>,
{
    let it: (usize, I) = iter_nodes.next().unwrap();
}

#[apply(Gat!)]
pub fn extend2<'a, 'b, I, L>(
    &mut self,
    iter_nodes: &'b mut L,
)
where
    I : Iterator<Item = usize>,
    L : LendingIterator,
 // <L as LendingIterator>::Item<'b>           = (usize, I) ,
    <L as LendingIterator>::Item<'b> : Is<EqTo = (usize, I)>,
{
    let it: (usize, I) = Is::eq_to(iter_nodes.next().unwrap());
}
```

  - <details><summary><code>trait Is</code> helper</summary>

	```rust
	trait Is {
	    type EqTo;
	
	    fn eq_to(this: Self) ->  Self::EqTo;
	}
	
	impl<T> Is for T {
	    type EqTo = Self;
	
	    fn eq_to(this: Self) ->  Self::EqTo {
	        this
	    }
	}
	```

	</details>

  - ### [Playground](https://www.rustexplorer.com/b/v4cep1)

**Comment by vigna** [comment]
That's definitely type magick. I understand that `Is::eq_to()` will extract the type, but I can't see how to write a `while let Some((x,i)) = iter.next() {}`. I mean, I tried, but...

[Playground](https://www.rustexplorer.com/b#LyoKW2RlcGVuZGVuY2llc10KbGVuZGluZy1pdGVyYXRvci52ZXJzaW9uID0gIjAuMS43IgojIG5vdWdhdC52ZXJzaW9uID0gIjAuMi40IgoqLwpmbiBtYWluKCkge30KCnVzZSA6OmxlbmRpbmdfaXRlcmF0b3I6OnByZWx1ZGU6Oio7CgpzdHJ1Y3QgRm9vOwoKdHJhaXQgSXMgewogICAgdHlwZSBFcVRvOwoKICAgIGZuIGVxX3RvKHRoaXM6IFNlbGYpIC0+IFNlbGY6OkVxVG87Cn0KCmltcGw8VD4gSXMgZm9yIFQgewogICAgdHlwZSBFcVRvID0gU2VsZjsKCiAgICBmbiBlcV90byh0aGlzOiBTZWxmKSAtPiBTZWxmOjpFcVRvIHsKICAgICAgICB0aGlzCiAgICB9Cn0KCmltcGwgRm9vIHsKICAgICNbYXBwbHkoR2F0ISldCiAgICBwdWIgZm4gZXh0ZW5kPCdhLCAnYiwgSSwgTD4oJm11dCBzZWxmLCBpdGVyX25vZGVzOiAmJ2IgbXV0IEwpCiAgICB3aGVyZQogICAgICAgIEk6IEl0ZXJhdG9yPEl0ZW0gPSB1c2l6ZT4sCiAgICAgICAgTDogZm9yPCdjPiBMZW5kaW5nSXRlcmF0b3I8SXRlbTwnYz4gPSAodXNpemUsIEkpPiwKICAgIHsKICAgICAgICBsZXQgaXQ6ICh1c2l6ZSwgSSkgPSBpdGVyX25vZGVzLm5leHQoKS51bndyYXAoKTsKICAgIH0KCiAgICAjW2FwcGx5KEdhdCEpXQogICAgcHViIGZuIGV4dGVuZDI8J2EsICdiLCBJLCBMPigmbXV0IHNlbGYsIGl0ZXJfbm9kZXM6ICYnYiBtdXQgTCkKICAgIHdoZXJlCiAgICAgICAgSTogSXRlcmF0b3I8SXRlbSA9IHVzaXplPiwKICAgICAgICBMOiBMZW5kaW5nSXRlcmF0b3IsCiAgICAgICAgLy8gPEwgYXMgTGVuZGluZ0l0ZXJhdG9yPjo6SXRlbTwnYj4gICAgICAgICAgID0gKHVzaXplLCBJKSAsCiAgICAgICAgPEwgYXMgTGVuZGluZ0l0ZXJhdG9yPjo6SXRlbTwnYj46IElzPEVxVG8gPSAodXNpemUsIEkpPiwKICAgIHsKICAgICAgICB3aGlsZSBsZXQgU29tZShpdGVtKSA9IGl0ZXJfbm9kZXMubmV4dCgpIHsKICAgICAgICAgICAgbGV0IF8gPSBJczo6ZXFfdG8oaXRlbSk7CiAgICAgICAgfQogICAgfQp9Cg==)

**Comment by danielhenrymantilla** [comment]
Yeah, this is a bit subtle to explain; do you happen to be on Discord or some other instant messaging platform? It may be easier to work with there than over here 🙂 

 1. The Rust Community Discord server: https://discord.gg/rust-lang-community

 1. You can then create a post over [`#rust-help-forum`](https://discord.com/channels/273534239310479360/981431291683700776)

 1. You can then ping me when doing so: `@yandros`

___

Otherwise, the TL,DR, _if we remove `usize` from the situation to simplify things_, what you have is (ignoring `nougat` syntax sugar/hacks, imagining full GATs all the way):

<details>

```rs
<L> …
where
	for<'c>
	    <L as LendingIterator>::Item<'c> : Iterator<Item = …>
	,
```

_i.e._,

```rs
<L>
where
	for<'c>
	    exists I<'c> : Iterator<Item = …>
	    where
	        L : LendingIterator<Item<'c> = I<'c>>
	    ,
```

(in pseudo-code).

But when we write:

```rs
<L, I>
where
    for<'c>
        L : LendingIterator<Item<'c> = I>
    ,
```

Notice how we are expressing that the lended item does not depend on the `'c = 'next` lending-lifetime of `fn next(&'c mut self)`! So that boils down to defeating the very purpose of `LendingIterator`s to begin with!

Hence my:

> either you write the bound as valid for all lifetimes (while this seems nice, in practice, **it will only be usable if the `LendingIterator` isn't truly lending / if `I` does not depend on `'c`**)

Now, when we do write:

```rs
<L as LendingIterator>::Item<'b>: Is<EqTo = I>
```

what we are saying is no longer a `for<'c>`-quantified clause, but, instead, a "punctual" statement/contraint/clause, stating that when the input `'c` happens to be exactly equal to `'b`, we are getting this outer/fixed `I` iterator.

  - This has the advantage of not _eagerly_ restricting the `for<'c> …::Item<'c>` to be some `'c`-independent fixed iterator `I`, as we've just seen in the previous part;

  - But it does restrict the `.next()` call(s) able to yield this known `I` to be of this fixed `'b` lifetime. In Rust, generic lifetime parameters (thus provided by the _caller_), necessarily span beyond the internal body of the callee's function:

	```rust
	    iterator.next() // ---- start of borrow ---+
	                                            // |
	                                            // |
	                                            // |
	} // <----------- 'fn body --------------------|
	                                            // ⋮ caller leeway in their choice of `'b`
	// <--------------- end of 'b -----------------+
	```

    But, if you want to call `.next()` multiple times inside that _callee_ function, you need the borrow of each `.next()` call (but for the last one) to be shorter than `'fn`:

	```rs
	    iterator.next() // ---- start of borrow ---+
	                                            // |
	                                            // |
	    // <-- first borrow must have ended here --+
	    
	    iterator.next() // - start of new borrow --+
	                                            // |
	                                            // |
	    // <--- snd borrow must have ended here ---+
	    
	    // etc.
	    
	} // <----------- 'fn body --------------------|
	```

    Hence the problem.

So, back to my `exists I<'c>` pseudo-code, what we wanted to express was:

```rs
<L>
where
	for<'c>
	    exists I<'c> : Iterator<Item = …>
	    where
	        L : LendingIterator<Item<'c> = I<'c>>
	    ,
```

Which, alas, is not something directly expressible in Rust. What we do, in these cases, is just to skip this `I<'c>` "alias", and always refer to `L::Item<'c>`:

```rs
<L>
where
    L : LendingIterator,
    for<'c>
        Item<'c, L> : Iterator<Item = …>
    ,
```

Now, the problem is that in your case you had `(usize, I<'c>)` rather than just `I<'c>`. This is where the current expressibility of Rust really runs into limitations.

 1. Since in this form we can only write trait bounds for `L::Item<'c>`,

 1. but since trait bounds allow introducing new associated items,

 1. then the solution would be to "trait-ify" being a `(T, U)` 2-tuple:

	```rs
	trait Tuple2 {
	    type _0;
	    type _1;
	
	    fn is_tuple(self) -> (Self::_0, Self::_1);
	}
	
	impl<T, U> Tuple2 for (T, U) {
	    type _0 = T;
	    type _1 = U;
	
	    fn is_tuple(self) -> (Self::_0, Self::_1) {
	        self
	    }
	}
	```

 1. And from there:

	```rs
    #[apply(Gat!)]
    pub fn extend<L>(&mut self, iter_nodes: &mut L)
    where
        L : LendingIterator,
        for<'c>
            Item<'c, L> : Tuple2<_0 = usize>
        ,
        for<'c>
            <Item<'c, L> as Tuple2>::_1 : Iterator<Item = usize>
        ,
    {
        while let Some((idx, iter)) = iter_nodes.next().map(|it| it.is_tuple()) {
            for n in iter {
                let _: usize = n;
            }
        }
    }
	```

[Playground](https://www.rustexplorer.com/b#LyoKW2RlcGVuZGVuY2llc10KbGVuZGluZy1pdGVyYXRvci52ZXJzaW9uID0gIjAuMS43IgojIG5vdWdhdC52ZXJzaW9uID0gIjAuMi40IgoqLwpmbiBtYWluKCkge30KCnVzZSA6OmxlbmRpbmdfaXRlcmF0b3I6OnByZWx1ZGU6Oio7CgpzdHJ1Y3QgRm9vOwoKdHJhaXQgSXMgewogICAgdHlwZSBFcVRvOwoKICAgIGZuIGVxX3RvKHRoaXM6IFNlbGYpIC0+IFNlbGY6OkVxVG87Cn0KCmltcGw8VD4gSXMgZm9yIFQgewogICAgdHlwZSBFcVRvID0gU2VsZjsKCiAgICBmbiBlcV90byh0aGlzOiBTZWxmKSAtPiBTZWxmOjpFcVRvIHsKICAgICAgICB0aGlzCiAgICB9Cn0KCnRyYWl0IFR1cGxlMiB7CiAgICB0eXBlIF8wOwogICAgdHlwZSBfMTsKCiAgICBmbiBpc190dXBsZShzZWxmKSAtPiAoU2VsZjo6XzAsIFNlbGY6Ol8xKTsKfQoKaW1wbDxULCBVPiBUdXBsZTIgZm9yIChULCBVKSB7CiAgICB0eXBlIF8wID0gVDsKICAgIHR5cGUgXzEgPSBVOwoKICAgIGZuIGlzX3R1cGxlKHNlbGYpIC0+IChTZWxmOjpfMCwgU2VsZjo6XzEpIHsKICAgICAgICBzZWxmCiAgICB9Cn0KCmltcGwgRm9vIHsKICAgICNbYXBwbHkoR2F0ISldCiAgICBwdWIgZm4gZXh0ZW5kPEw+KCZtdXQgc2VsZiwgaXRlcl9ub2RlczogJm11dCBMKQogICAgd2hlcmUKICAgICAgICBMIDogTGVuZGluZ0l0ZXJhdG9yLAogICAgICAgIGZvcjwnYz4KICAgICAgICAgICAgPEwgYXMgTGVuZGluZ0l0ZXJhdG9yPjo6SXRlbTwnYz4gOiBUdXBsZTI8XzAgPSB1c2l6ZT4KICAgICAgICAsCiAgICAgICAgZm9yPCdjPgogICAgICAgICAgICA8PEwgYXMgTGVuZGluZ0l0ZXJhdG9yPjo6SXRlbTwnYz4gYXMgVHVwbGUyPjo6XzEgOiBJdGVyYXRvcjxJdGVtID0gdXNpemU+CiAgICAgICAgLAogICAgewogICAgICAgIHdoaWxlIGxldCBTb21lKChpZHgsIGl0ZXIpKSA9IGl0ZXJfbm9kZXMubmV4dCgpLm1hcCh8aXR8IGl0LmlzX3R1cGxlKCkpIHsKICAgICAgICAgICAgZm9yIG4gaW4gaXRlciB7CiAgICAgICAgICAgICAgICBsZXQgXzogdXNpemUgPSBuOwogICAgICAgICAgICB9CiAgICAgICAgfQogICAgfQp9Cg==)

FWIW, there is kind of a trick/hack in rust wherein you can use _nested_ `impl Trait`, but only when within the initial `<….>` generics rather than `where` clauses. This can be used here to alleviate a bit the syntax:

  - EDIT: ⚠️ beware of this hack ⚠️ The nested `impl Trait` syntax seems to introduce an anonymous _but `for<>`-lifetime-agnostic_ parameter to describe it, thereby introducing fixed lifetime constraints which can lead to `higher rank lifetime error`s

```rs
#[apply(Gat!)]
pub fn extend<
    L : for<'c> LendingIterator<
        // this would be our `(usize, I<'c>)`
        Item<'c> = impl Tuple2<               // (
            _0 = usize,                       //     usize,
            _1 = impl Iterator<Item = usize>, //     I<'c>,
        >,                                    // )
    >,
>(
    &mut self,
    iter_nodes: &mut L,
)
{
    while let Some((idx, iter)) = iter_nodes.next().map(|it| it.is_tuple()) {
        for n in iter {
            let _: usize = n;
        }
    }
}
```

  - [Playground](https://www.rustexplorer.com/b#LyoKW2RlcGVuZGVuY2llc10KbGVuZGluZy1pdGVyYXRvci52ZXJzaW9uID0gIjAuMS43IgojIG5vdWdhdC52ZXJzaW9uID0gIjAuMi40IgoqLwpmbiBtYWluKCkge30KCnVzZSA6OmxlbmRpbmdfaXRlcmF0b3I6OnByZWx1ZGU6Oio7CgpzdHJ1Y3QgRm9vOwoKdHJhaXQgSXMgewogICAgdHlwZSBFcVRvOwoKICAgIGZuIGVxX3RvKHRoaXM6IFNlbGYpIC0+IFNlbGY6OkVxVG87Cn0KCmltcGw8VD4gSXMgZm9yIFQgewogICAgdHlwZSBFcVRvID0gU2VsZjsKCiAgICBmbiBlcV90byh0aGlzOiBTZWxmKSAtPiBTZWxmOjpFcVRvIHsKICAgICAgICB0aGlzCiAgICB9Cn0KCnRyYWl0IFR1cGxlMiB7CiAgICB0eXBlIF8wOwogICAgdHlwZSBfMTsKCiAgICBmbiBpc190dXBsZShzZWxmKSAtPiAoU2VsZjo6XzAsIFNlbGY6Ol8xKTsKfQoKaW1wbDxULCBVPiBUdXBsZTIgZm9yIChULCBVKSB7CiAgICB0eXBlIF8wID0gVDsKICAgIHR5cGUgXzEgPSBVOwoKICAgIGZuIGlzX3R1cGxlKHNlbGYpIC0+IChTZWxmOjpfMCwgU2VsZjo6XzEpIHsKICAgICAgICBzZWxmCiAgICB9Cn0KCmltcGwgRm9vIHsKICAgICNbYXBwbHkoR2F0ISldCiAgICBwdWIgZm4gZXh0ZW5kPAogICAgICAgIEwgOiBmb3I8J2M+IExlbmRpbmdJdGVyYXRvcjwKICAgICAgICAgICAgLy8gdGhpcyB3b3VsZCBiZSBvdXIgYCh1c2l6ZSwgSTwnYz4pYAogICAgICAgICAgICBJdGVtPCdjPiA9IGltcGwgVHVwbGUyPCAgICAgICAgICAgICAgIC8vICgKICAgICAgICAgICAgICAgIF8wID0gdXNpemUsICAgICAgICAgICAgICAgICAgICAgICAvLyAgICAgdXNpemUsCiAgICAgICAgICAgICAgICBfMSA9IGltcGwgSXRlcmF0b3I8SXRlbSA9IHVzaXplPiwgLy8gICAgIEk8J2M+LAogICAgICAgICAgICA+LCAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgIC8vICkKICAgICAgICA+LAogICAgPigKICAgICAgICAmbXV0IHNlbGYsCiAgICAgICAgaXRlcl9ub2RlczogJm11dCBMLAogICAgKQogICAgewogICAgICAgIHdoaWxlIGxldCBTb21lKChpZHgsIGl0ZXIpKSA9IGl0ZXJfbm9kZXMubmV4dCgpLm1hcCh8aXR8IGl0LmlzX3R1cGxlKCkpIHsKICAgICAgICAgICAgZm9yIG4gaW4gaXRlciB7CiAgICAgICAgICAgICAgICBsZXQgXzogdXNpemUgPSBuOwogICAgICAgICAgICB9CiAgICAgICAgfQogICAgfQp9Cg==)
    
</details>

**Comment by vigna** [comment]
> Yeah, this is a bit subtle to explain; do you happen to be on Discord or some other instant messaging platform? It may be easier to work with there than over here 🙂
>

I think I'm bothering you more than enough here!

So, thank you a lot for the explanation: turns out I was desperately trying to do what you suggest, but I was looking for Rust syntax that would make it possible to extract the type of a component of a tuple, and I couldn't find any. I didn't think about the trait—that's a bit too far for me.

I've implemented the idea and it seems to work even with actual GATs (using gat_lending_iterator), except that when I try to call a function with the multi-line parameter declaration above using a `LendingIterator<Item<'a> = (usize, Self::Successors<'a>)>` I get 
```
the trait bound `for<'c> <_ as gat_lending_iterator::LendingIterator>::Item<'c>: traits::graph::Tuple2` is not satisfied
the trait `traits::graph::Tuple2` is implemented for `(T, U)`
```
which puzzles me because the item is a 2-tuple. That might be a limitation of the compiler that does not apply to your HKT implementation, but it seems to be something else. I tried to reproduce the situation using your library but I got stuck trying to set the Item type to a pair (possibly I misunderstood the `Item<'_,L>` syntax).

Basically, I need to return a  `LendingIterator<Item<'a> = (usize, Self::Successors<'a>)>` from a function and pass it to another. With the complex signature you propose the function finally compiles, but then it doesn't accept what I would expect as a possible argument with the error above.

If you can share some more wisdom I would be immensely grateful!

**Comment by vigna** [comment]
And, I was able to make this into a [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e2d97fda42c92f9509691e6da2c636be).

**Comment by danielhenrymantilla** [comment]
The following [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=701d8267864fedfa8a74654edb49ad16) compiles

**Comment by vigna** [comment]
I see—you really need a different definition of lending iterator to make this work.

I realize it is similar to the type structure of your library, so probably the `LendingIterator` definition we were using was wrong from the start. Thank you again, I think we'll be able to work out this idea.

**Comment by vigna** [comment]
So, in the end this was the idea we were looking for. With a bit of `'static` we have been able to make everything work as we were expecting. I [credited your contribution in the docs](https://github.com/vigna/webgraph-rs/blob/havoc/src/traits/iter.rs).
